src/util.rs

```diff
@@ -37,57 +40,43 @@ pub fn vector_with_length<F: FieldElement>(len: usize) -> Vec<F> {
 }
 
 /// Unpacked proof with subcomponents
-#[derive(Debug)]
-pub struct UnpackedProof<'a, F: FieldElement> {
+#[derive(Debug, Serialize, Deserialize)]
+#[serde(bound(deserialize = ""))]
+pub struct UnpackedProof<F: FieldElement> {
     /// Data
-    pub data: &'a [F],
+    pub data: Vec<F>,
     /// Zeroth coefficient of polynomial f
-    pub f0: &'a F,
+    pub f0: F,
     /// Zeroth coefficient of polynomial g
-    pub g0: &'a F,
+    pub g0: F,
     /// Zeroth coefficient of polynomial h
-    pub h0: &'a F,
+    pub h0: F,
     /// Non-zero points of polynomial h
-    pub points_h_packed: &'a [F],
+    pub points_h_packed: Vec<F>,
 }
 
 /// Unpacked proof with mutable subcomponents
-#[derive(Debug)]
-pub struct UnpackedProofMut<'a, F: FieldElement> {
+#[derive(Debug, Serialize, Deserialize)]
+#[serde(bound(deserialize = ""))]
+pub struct UnpackedProofMut<F: FieldElement> {
     /// Data
-    pub data: &'a mut [F],
+    pub data: Vec<F>,
     /// Zeroth coefficient of polynomial f
-    pub f0: &'a mut F,
+    pub f0: F,
     /// Zeroth coefficient of polynomial g
-    pub g0: &'a mut F,
+    pub g0: F,
     /// Zeroth coefficient of polynomial h
-    pub h0: &'a mut F,
+    pub h0: F,
     /// Non-zero points of polynomial h
-    pub points_h_packed: &'a mut [F],
+    pub points_h_packed: Vec<F>,
 }
 
 /// Unpacks the proof vector into subcomponents
 pub(crate) fn unpack_proof<F: FieldElement>(
     proof: &[F],
-    dimension: usize,
-) -> Result<UnpackedProof<F>, SerializeError> {
-    // check the proof length
-    if proof.len() != proof_length(dimension) {
-        return Err(SerializeError::UnpackInputSizeMismatch);
-    }
-    // split share into components
-    let (data, rest) = proof.split_at(dimension);
-    if let ([f0, g0, h0], points_h_packed) = rest.split_at(3) {
-        Ok(UnpackedProof {
-            data,
-            f0,
-            g0,
-            h0,
-            points_h_packed,
-        })
-    } else {
-        Err(SerializeError::UnpackInputSizeMismatch)
-    }
+) -> Result<UnpackedProof<F>, Box<bincode::ErrorKind>> {
+    let bytes = bincode::serialize(proof).unwrap();
```

Comment:
It's less than ideal to serialize the FieldElements into bytes (since they may take up a different amount of space depending on the field size), and then to deserialize into the unpacked structure.